### PR TITLE
Add attribute flag building and parsing

### PR DESCRIPTION
**Pull request by jefferai**
This adds a few command sketches but that's not the main point of the PR
and filling those out will come in a different PR. This adds a new flag
type that can combine inputs to a single target slice, retaining the
ability to distinguish where it came from but also keep ordering across
different flag names. It also implements a function that does the
heuristic parsing with fallback.

### CODE REVIEW

File: internal/cmd/common/flags_test.go [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+				{
+					Name:  "attr",
+					Keys:  []string{"baz"},
+					Value: "9820",
+				},
+			},
+		},
+		{
+			name: "mixed-segments",
+			args: []string{"-num-attr", "foo.bar.baz=9820", "-string-attr", "bar.baz.foo=9820", "-attr", "baz.foo.bar=9820"},
```
[the remark this answers is not in the document]
Comment (jefferai) [reply internal/cmd/common/flags_test.go:94]: Building these into interface slices is actually handled in a different function and thus tested in a different test. The purpose of this subtest is purely to ensure that the splitting logic works as expected.

[the remark this answers is not in the document]
Comment (jefferai) [reply internal/cmd/common/flags_test.go:94]: (See the two tests starting at https://github.com/hashicorp/boundary/pull/1562/files/12762323edeb87750ad2001aafc3e33ee11dbb4a#diff-e8569cddb838d7e07805feff0fc49f3c12482dc9297d810b27d7addc4cffccf0R260)

File: internal/cmd/commands/hostcatalogscmd/plugin_funcs.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+func (c *PluginCommand) extraPluginHelpFunc(helpMap map[string]func() string) string {
+	var helpStr string
+	switch c.Func {
+	case "create":
+		helpStr = base.WrapForHelpText([]string{
+			"Usage: boundary host-catalogs create plugin [options] [args]",
+			"",
+			"  Create a plugin-type host catalog. Example:",
+			"",
+			`    $ boundary host-catalogs create plugin -name prodops -description "Plugin host-catalog for ProdOps"`,
```
[the remark this answers is not in the document]
Comment (jefferai) [reply internal/cmd/commands/hostcatalogscmd/plugin_funcs.go:16]: Added

File: internal/cmd/commands/hostcatalogscmd/plugin_funcs.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+			"",
+		})
+
+	case "update":
+		helpStr = base.WrapForHelpText([]string{
+			"Usage: boundary host-catalogs update plugin [options] [args]",
+			"",
+			"  Update a plugin-type host catalog given its ID. Example:",
+			"",
+			`    $ boundary host-catalogs update plugin -id hcst_1234567890 -name "devops" -description "Plugin host-catalog for DevOps"`,
```
[the remark this answers is not in the document]
Comment (jefferai) [reply internal/cmd/commands/hostcatalogscmd/plugin_funcs.go:27]: I'll wait on this until we actually switch static catalogs over to generating non-subtyped IDs, which I think is going to be the release after the plugins come out.

File: internal/cmd/common/flags_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jefferai) [reply internal/cmd/common/flags_test.go:94]: Resolving as I believe the referenced tests have the functionality in question

File: internal/cmd/common/flags.go [hunk 4]
```diff
@@ -93,3 +100,240 @@ func PopulateCommonFlags(c *base.Command, f *base.FlagSet, resourceType string,
 		}
 	}
 }
+
+func PopulateAttributeFlags(c *base.Command, f *base.FlagSet, flagNames map[string][]string, command string) {
```
Comment (jimlambrt) [inline internal/cmd/common/flags.go:104]: It might be nice to have a godoc for this exported func?

File: internal/cmd/common/flags.go [hunk 5]
```diff
@@ -93,3 +100,240 @@ func PopulateCommonFlags(c *base.Command, f *base.FlagSet, resourceType string,
...
+	keyDelimiter := "."
+	for _, name := range flagNames[command] {
+		switch name {
+		case "attributes":
+			f.StringVar(&base.StringVar{
+				Name:   "attributes",
+				Target: &c.FlagAttributes,
+				Usage:  `A JSON map value to use as the entirety of the request's attribute map. Usually this will be soured from a file via "file://" syntax.`,
```
Comment (jimlambrt) [inline internal/cmd/common/flags.go:112]: soured == sourced?

File: internal/cmd/base/flags.go [hunk 6]
```diff
@@ -748,3 +751,118 @@ func (f *FlagSet) Var(value flag.Value, name, usage string) {
...
+	Usage          string
+	Hidden         bool
+	Target         *[]CombinedSliceFlagValue
+	Completion     complete.Predictor
+	KvSplit        bool
+	KeyDelimiter   *string
+	ProtoCompatKey bool
+}
+
+func (f *FlagSet) CombinationSliceVar(i *CombinationSliceVar) {
```
Comment (jimlambrt) [inline internal/cmd/base/flags.go:782]: might be nice to have a godoc for this.

File: internal/cmd/base/flags.go [hunk 7]
```diff
@@ -748,3 +751,118 @@ func (f *FlagSet) Var(value flag.Value, name, usage string) {
...
+		kvSplit:        kvSplit,
+		keyDelimiter:   keyDelimiter,
+		protoCompatKey: protoCompatKey,
+		target:         target,
+	}
+}
+
+var protoIdentifierRegex = regexp.MustCompile("^[a-zA-Z][A-Za-z0-9_]*$")
+
+func (c *combinedSliceValue) Set(val string) error {
```
Comment (jimlambrt) [inline internal/cmd/base/flags.go:822]: godoc?

**Review by jimlambrt — COMMENTED** [review]
Just a few small things.

File: internal/cmd/commands/hostcatalogscmd/plugin_funcs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/cmd/commands/hostcatalogscmd/plugin_funcs.go]: But this is the helper text for plugin based catalogs, not static.

File: internal/cmd/common/flags_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/cmd/common/flags_test.go:94]: The behavior I was trying to call out to be tested was an order preserving behavior when jumping between -num-attr and -attr for the same key.  I don't see that behavior captured anywhere I don't think.

**Pull request merged**